README.md
# Flutter Awesome Bottom Sheet

## Features

* Custom child widgets
* Adaptive design

## Supported platforms

* Flutter Android
* Flutter iOS
* Flutter web
* Flutter desktop
* Flutter macOS
* Flutter linux

## Installation

Add ```flutter_awesome_bottom_sheet: <later-version>``` to your `pubspec.yaml` dependencies. And import it:

```
import 'package:flutter_awesome_bottom_sheet/flutter_awesome_bottom_sheet.dart';
```

#### How to Use ####

## How to Use ##
create object
```
final AwesomeBottomSheet _awesomeBottomSheet = AwesomeBottomSheet();
```

## Example

Simple Awesome Sheet

```          
 _awesomeBottomSheet.show(
      context: context,
      title: const Text("Lorem Ipsum"),
      description: const Text("Flutter Awesome Bottom sheet content here..."),
      color: CustomSheetColor(
        mainColor: const Color(0xff2979FF),
        accentColor: const Color(0xff0D47A1),
        iconColor: Colors.white,
      ),
      positive: AwesomeSheetAction(
        onPressed: () {
          Navigator.of(context).pop();
        },
        title: 'CANCEL',
      ),
    );
```

## Add Buttons

```
 _awesomeBottomSheet.show(
     ...
      positive: AwesomeSheetAction(
        onPressed: () {
          Navigator.of(context).pop();
        },
        title: 'DELETE',
      ),
      negative: AwesomeSheetAction(
        onPressed: () {
          Navigator.of(context).pop();
        },
        title: 'CANCEL',
      ),
    );
```

Add Icons in content and action button

```

 _awesomeBottomSheet.show(
     ...
      icon: Icons.signal_cellular_connected_no_internet_4_bar,
      positive: AwesomeSheetAction(
        onPressed: () {
          Navigator.of(context).pop();
        },
        title: 'OPEN SETTING',
        icon: Icons.settings,
      ),
      negative: AwesomeSheetAction(
        onPressed: () {
          Navigator.of(context).pop();
        },
        title: 'CANCEL',
      ),
    );

```

## Custom color

```
 _awesomeBottomSheet.show(
      ...
      color: CustomSheetColor(
        mainColor: Colors.white,
        accentColor: const Color(0xff5A67D8),
        iconColor: const Color(0xff5A67D8),
      ),
      ...
    ); 
```

## Screenshot
|             Simple             |             Action             |             Icon             |
| :----------------------------: | :----------------------------: | :-------------------: |
| ![simple](simple.png)          | ![action](action.png)          | ![icon](icon.png) |



CHANGELOG.md
## 0.0.1

* Initial release.
* 
## 0.0.1

* Minor changes.
* Update design for better experience


test/flutter_awesome_bottom_sheet_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_awesome_bottom_sheet/flutter_awesome_bottom_sheet.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/awesome_bottom_sheet.dart
import 'package:flutter/material.dart';

/// Custom color class for multiple theme color
class CustomSheetColor {
  Color mainColor;
  Color accentColor;
  Color iconColor;

  CustomSheetColor(
      {required this.mainColor,
      required this.accentColor,
      required this.iconColor});
}

class AwesomeBottomSheet {
  show({
    required BuildContext context,
    Text? title,
    required Text description,
    required CustomSheetColor color,
    required AwesomeSheetAction positive,
    EdgeInsets? contentPadding =
        const EdgeInsets.symmetric(horizontal: 24.0, vertical: 24.0),
    EdgeInsets? actionPadding = const EdgeInsets.all(8.0),
    AwesomeSheetAction? negative,
    IconData? icon,
    double iconSize = 52,
    double titleFontSize = 24,
    double descriptionFontSize = 18,
    String fontFamily = 'circular',
    bool useRootNavigator = false,
    bool isDismissible = true,
    double radiusTopLeft = 10,
    double radiusTopRight = 10,
    double radiusBottomLeft = 0,
    double radiusBottomRight = 0,
  }) {
    showModalBottomSheet(
      isDismissible: isDismissible,
      enableDrag: isDismissible,
      context: context,
      isScrollControlled: true,
      backgroundColor: Colors.transparent,
      useRootNavigator: useRootNavigator,
      builder: (BuildContext context) {
        return ClipRRect(
          borderRadius: BorderRadius.only(
            topLeft: Radius.circular(radiusTopLeft),
            topRight: Radius.circular(radiusTopRight),
            bottomLeft: Radius.circular(radiusBottomLeft),
            bottomRight: Radius.circular(radiusBottomRight),
          ),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Container(
                width: double.infinity,
                color: color.mainColor,
                padding: contentPadding,
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    title == null
                        ? Container()
                        : DefaultTextStyle(
                            style: TextStyle(
                              fontSize: titleFontSize,
                              color: Colors.white,
                              fontWeight: FontWeight.bold,
                              fontFamily: fontFamily,
                            ),
                            textAlign: TextAlign.start,
                            child: title),
                    _buildContent(color, description, icon, fontFamily,
                        descriptionFontSize, iconSize)
                  ],
                ),
              ),
              Container(
                padding: actionPadding,
                color: color.accentColor,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.end,
                  children: _buildActions(positive, negative),
                ),
              )
            ],
          ),
        );
      },
    );
  }

  /// return content widget
  _buildContent(CustomSheetColor color, Text description, IconData? icon,
      String fontFamily, double descriptionFontSize, double iconSize) {
    return Padding(
      padding: const EdgeInsets.only(top: 12.0),
      child: SingleChildScrollView(
        child: icon != null
            ? Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: <Widget>[
                  Expanded(
                    child: DefaultTextStyle(
                        style: TextStyle(
                            fontSize: descriptionFontSize,
                            color: Colors.white,
                            fontFamily: fontFamily),
                        child: description),
                  ),
                  const SizedBox(
                    width: 16,
                  ),
                  Icon(
                    icon,
                    size: iconSize,
                    color: color.iconColor,
                  )
                ],
              )
            : DefaultTextStyle(
                style: TextStyle(
                  fontSize: descriptionFontSize,
                  fontFamily: fontFamily,
                  color: Colors.white,
                ),
                child: description,
              ),
      ),
    );
  }

  /// action widget
  _buildActions(AwesomeSheetAction? positive, AwesomeSheetAction? negative) {
    List<AwesomeSheetAction> actions = [];

    // This order is important
    // It helps to place the positive at the right and the negative before
    if (negative != null) {
      actions.add(negative);
    }

    if (positive != null) {
      actions.add(positive);
    }

    return actions;
  }
}

/// action button widget

class AwesomeSheetAction extends StatelessWidget {
  final String title;
  final VoidCallback onPressed;
  final IconData? icon;
  final Color color;
  VoidCallback? onLongPress;

  AwesomeSheetAction({
    Key? key,
    required this.title,
    required this.onPressed,
    this.onLongPress,
    this.icon,
    this.color = Colors.white,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return icon == null
        ? TextButton(
            onPressed: onPressed,
            onLongPress: onLongPress,
            child: Text(
              title,
              style: TextStyle(
                color: color,
              ),
            ),
          )
        : TextButton.icon(
            onPressed: onPressed,
            onLongPress: onLongPress,
            label: Text(
              title,
              style: TextStyle(
                color: color,
              ),
            ),
            icon: Icon(
              icon,
              color: color,
            ),
          );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_awesome_bottom_sheet/flutter_awesome_bottom_sheet.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Awesome Bottom Sheet',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
          brightness: Brightness.dark,
          primarySwatch: Colors.blue,
          useMaterial3: true),
      home: const HomePage(title: 'Awesome Bottom Sheet'),
    );
  }
}

class HomePage extends StatefulWidget {
  final String title;
  const HomePage({Key? key,required this.title}) : super(key: key);

  @override
  HomePageState createState() => HomePageState();
}

class HomePageState extends State<HomePage> {
  ///Object declaration
  final AwesomeBottomSheet _awesomeBottomSheet = AwesomeBottomSheet();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      backgroundColor: Colors.white,
      body: SizedBox(
        child: ListView(
          children: <Widget>[
            const SizedBox(
              height: 10,
            ),
            ///Simple Awesome Sheet
            BuildButton(
              context: context,
              text: 'Simple Awesome Sheet',
              onClick: () {
                _awesomeBottomSheet.show(
                  context: context,
                  title: const Text("Lorem Ipsum"),
                  description: const Text(
                      "Flutter Awesome Bottom sheet content here..."),
                  color: CustomSheetColor(
                    mainColor: const Color(0xff2979FF),
                    accentColor: const Color(0xff0D47A1),
                    iconColor: Colors.white,
                  ),
                  positive: AwesomeSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),

            ///Action Awesome Sheet
            BuildButton(
              context: context,
              text: 'Action Awesome Sheet',
              onClick: () {
                _awesomeBottomSheet.show(
                  isDismissible: false,
                  context: context,
                  title: const Text("Delete this post?"),
                  description: const Text(
                      "This action will permanently delete this post."),
                  color: CustomSheetColor(
                    mainColor: const Color(0xfff33838),
                    accentColor: const Color(0xffab1d1d),
                    iconColor: Colors.white,
                  ),
                  icon: Icons.delete,
                  positive: AwesomeSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'DELETE',
                  ),
                  negative: AwesomeSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),


            ///Icon Awesome Sheet

            BuildButton(
              context: context,
              text: 'Icon Awesome Sheet',
              onClick: () {
                _awesomeBottomSheet.show(
                  context: context,
                  title: const Text("Attention"),
                  description: const Text(
                      'Your app is not connected to internet actually, please turn on Wifi/Cellular data.'),
                  color: CustomSheetColor(
                    mainColor: const Color(0xff2979FF),
                    accentColor: const Color(0xff0D47A1),
                    iconColor: Colors.white,
                  ),
                  icon: Icons.signal_cellular_connected_no_internet_4_bar,
                  positive: AwesomeSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'OPEN SETTING',
                    icon: Icons.settings,
                  ),
                  negative: AwesomeSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}

///Create Button widget for multiple usage
class BuildButton extends StatelessWidget {
  final BuildContext context;
  final String text;
  final VoidCallback onClick;

  const BuildButton({
    Key? key,
    required this.context,
    required this.text,
    required this.onClick,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.only(bottom: 8.0),
      margin: const EdgeInsets.symmetric(horizontal: 16),
      child: GestureDetector(
        onTap: onClick,
        child: Container(
          padding: const EdgeInsets.all(10),
          color: Colors.blue,
          child: SizedBox(
            width: double.infinity,
            child: Text(
              text,
              textAlign: TextAlign.center,
            ),
          ),
        ),
      ),
    );
  }
}
